### PR TITLE
Flowers: Optimize mushroom spread

### PR DESCRIPTION
**Pull request by tenplus1**
This pull splits the mushroom abm into a separate flowers.mushroom_spread(pos, node) function which is more optimized and stops shrooms spreading through solid walls as much.

### CODE REVIEW

File: mods/flowers/init.lua [hunk 1]
```diff
@@ -204,38 +204,34 @@ minetest.register_node("flowers:mushroom_brown", {
...
 -- Mushroom spread and death
 
+function flowers.mushroom_spread(pos, node)
+	if minetest.get_node_light(pos, nil) == 15 then
+		minetest.remove_node(pos)
+		return
+	end
+	local positions = minetest.find_nodes_in_area_under_air(
+		{x = pos.x - 1, y = pos.y - 2, z = pos.z - 1},
+		{x = pos.x + 1, y = pos.y + 1, z = pos.z + 1},
```
Comment (SmallJoker) [inline mods/flowers/init.lua]: Previously this was the range of `-2 ... 2`, why the change?

Comment (tenplus1) [inline mods/flowers/init.lua]: To specifically stop it spreading through solid walls, at least with 1 radius it an only spread through diagonals which are rare,

Comment (tenplus1) [inline mods/flowers/init.lua]: ...also mushrooms can spread through dirt so why not to the next available dirt block beside shroom :)

File: mods/flowers/init.lua [hunk 2]
```diff
@@ -204,38 +204,34 @@ minetest.register_node("flowers:mushroom_brown", {
...
+		{"group:soil", "group:tree"})
+	if #positions == 0 then
+		return
+	end
+	local pos2 = positions[math.random(#positions)]
+	pos2.y = pos2.y + 1
+	if minetest.get_node_light(pos, 0.5) < 4 and
```
Comment (paramat) [inline mods/flowers/init.lua]: <=3 is clearer (and no slower) because it makes clear that 3 is the maximum light level mushrooms can tolerate, this is useful information to know and more useful than knowing what level kills nushrooms.

**Comment by paramat** [comment]
This is actually slower code because it searches a volume instead of picking a random node. However it is much more likely to result in a mushroom spread, so the ABM needs to be tuned to act less often to keep the same spawning rate. A less often occuring ABM is a good thing so this will be a useful PR once tuned.

**Comment by paramat** [comment]
The old code had a 1 in 3 chance of choosing a surface node because the volume is xyz 5x3x5, so multiply the ABM 'chance' by 3.

**Comment by tenplus1** [comment]
@paramat - changes added.

**Comment by paramat** [comment]
Ok sorry, i just realised that the individual map accesses may have been slower than a bulk engine search.

File: mods/flowers/init.lua (hunk 1 above)
Comment (paramat) [inline mods/flowers/init.lua]: I'm not sure about the pos.y - 2 though, i know it makes sense for spread dowmwards through soil to an iair gap under soil, but that's an unlikely environment and the distance seems inconsistent.

Comment (paramat) [inline mods/flowers/init.lua]: I like the code, +1 if you change this to pos.y - 1.

**Comment by paramat** [comment]
Good improvement, +1 with a small change.

File: mods/flowers/init.lua (hunk 1 above)
Comment (paramat) [inline mods/flowers/init.lua]: Range being -1 to +1 in X Z is fine for me, it will improve performance.

**Comment by paramat** [comment]
:+1: 
SmallJoker approves http://irc.minetest.net/minetest-dev/2017-08-25#i_5054829 and is fine with -2 so i'll go with that.

**Comment by paramat** [comment]
bcf98df5fac3eb59b02d9fdb066ab92c6228d787
